crypto/cipher_extra/aead_test.cc

```diff
@@ -271,6 +271,203 @@ TEST_P(PerAEADTest, TestVector) {
+
+    {"AES_128_CBC_SHA256_TLS", EVP_aes_128_cbc_hmac_sha256,
+     "aes_128_cbc_sha256_tls_stitch_tests.txt",
+     kLimitedImplementation | RequiresADLength(EVP_AEAD_TLS1_AAD_LEN)},
+
+    {"AES_128_CBC_SHA256_TLS_IMPLICIT_IV", EVP_aes_128_cbc_hmac_sha256,
+     "aes_128_cbc_sha256_tls_stitch_implicit_iv_tests.txt",
+     kLimitedImplementation | RequiresADLength(EVP_AEAD_TLS1_AAD_LEN)},
+
+    {"AES_256_CBC_SHA1_TLS", EVP_aes_256_cbc_hmac_sha1,
+     "aes_256_cbc_sha1_tls_stitch_tests.txt",
+     kLimitedImplementation | RequiresADLength(EVP_AEAD_TLS1_AAD_LEN)},
+
+     {"AES_256_CBC_SHA1_TLS_IMPLICIT_IV", EVP_aes_256_cbc_hmac_sha1,
+     "aes_256_cbc_sha1_tls_stitch_implicit_iv_tests.txt",
+     kLimitedImplementation | RequiresADLength(EVP_AEAD_TLS1_AAD_LEN)},
+
+    {"AES_256_CBC_SHA256_TLS", EVP_aes_256_cbc_hmac_sha256,
+     "aes_256_cbc_sha256_tls_stitch_tests.txt",
+     kLimitedImplementation | RequiresADLength(EVP_AEAD_TLS1_AAD_LEN)},
+
+    {"AES_256_CBC_SHA256_TLS_IMPLICIT_IV", EVP_aes_256_cbc_hmac_sha256,
+     "aes_256_cbc_sha256_tls_stitch_implicit_iv_tests.txt",
+     kLimitedImplementation | RequiresADLength(EVP_AEAD_TLS1_AAD_LEN)},
+};
+
+class PerTLSLegacyAEADTest : public testing::TestWithParam<KnownTLSLegacyAEAD> {
+ public:
+  const EVP_CIPHER *legacy_aead() { return GetParam().func(); }
+};
+
+INSTANTIATE_TEST_SUITE_P(All, PerTLSLegacyAEADTest, testing::ValuesIn(kTLSLegacyAEADs),
+                         [](const testing::TestParamInfo<KnownTLSLegacyAEAD> &params)
+                             -> std::string { return params.param.name; });
+
+static void set_MAC_key(EVP_CIPHER_CTX *ctx, uint8_t *key, size_t mac_key_size) {
+  // In each TLS session, only need to set EVP_CTRL_AEAD_SET_MAC_KEY once.
+  ASSERT_TRUE(EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_MAC_KEY, mac_key_size, key))
+    << "EVP_CTRL_AEAD_SET_MAC_KEY failed: "
+    << ERR_reason_error_string(ERR_get_error());
+}
+
+static void set_TLS1_AAD(EVP_CIPHER_CTX *ctx, uint8_t *ad) {
+  // In each TLS session, ad should be set before each read/write operation because it includes sequence_num.
+  // Below ctrl returns hmac|pad len for encryption.
+  // For decryption, it returns digest size.
+  EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_TLS1_AAD, EVP_AEAD_TLS1_AAD_LEN, ad);
+}
+
+// |EVP_aes_128/256_cbc_hmac_sha1/256| tests.
+// Tests a TLS specific legacy AEAD against a series of test vectors from a file, using the
+// FileTest format. As an example, here's a valid test case:
+//
+// # DIGEST: 918c0df73de553b5bdffb7365f93a430292f6eea
+// KEY: 2993a340b9b3c589c7481df3f4183aa23fd8d7efd88503f78b8ed1c8e9ba2fd6773e0d0c
+// NONCE: 302a5f47e037446f5891d77df660ed82
+// IN: 302a5f47e037446f5891d77df660ed8286d641b877
+// AD: 97b684e0fb56f97c3903020015
+// CT: 6854710b76c79c9be84b3669fc4f05b17e67a9cc14
+// TAG: 9c6998bf3b172670c5f8613f479641468bbed4c68d093940c92de4
+// TAG_LEN: 20
+TEST_P(PerTLSLegacyAEADTest, TestVector) {
```

Comment:
The streaming APIs (e.g. `EVP_CipherUpdate`) doesn't work with stitched, right?

Reply:
I believe no. Besides, the stitch API itself is TLS specific, not generic cipher IMO.

Reply:
We should document how to use the stitched API carefully. It is easy to get wrong.

Reply:
Added.